Feat/#8 유저 커뮤니티 프로필 설정 기능

haeunsong opened the pull request:
- 닉네임 수정
- 닉네임 중복 테스트
- 역할 수정
- 회원가입 진행 상태 확인

프로필 사진 등록은 구현X. 회원가입 진행 상태 확인시 기본 false 로 진행.

haeunsong committed:
Feat(#8): user 닉네임/프로필사진/역할 선택 기능 추가

haeunsong committed:
Feat(#8): user 닉네임/프로필사진/역할 선택 기능 수정

haeunsong committed:
Feat(#8): nickname 중복 체크 로직 추가

haeunsong committed:
Test(#8): 기본 단위테스트 추가

jihoo2002 commented on src/main/java/com/keodam/keodam_backend/app/user/controller/UserController.java:
```diff
@@ -0,0 +1,93 @@
+    @PatchMapping("/nickname")
+    public ResponseEntity<ApiResponse<UserResponseDto>> updateNickname(
+            @RequestBody NicknameRequestDto request,
+            @AuthenticationPrincipal OAuth2User oAuth2User) {
```
`Authentication` 대신 `@AuthenticationPrincipal` 쓴 이유가 있으신가요? 궁금해요!

haeunsong replied:
이미 인증된 사용자 객체를 다룰 때에는 @AuthenticationPrincipal 를 쓰는게 더 간편하다고 하네요! Authentication 객체에서 principal(로그인한 사용자 객체) 만 뽑아주는게 @AuthenticationPrincipal 인것으로 이해하고 있습니당

jihoo2002 commented on src/main/java/com/keodam/keodam_backend/app/user/controller/UserController.java:
```diff
@@ -0,0 +1,93 @@
+        return ResponseEntity.ok(ApiResponse.onSuccess(response));
+    }
+
+
```
필요없는 개행이 있는 거 같아욥

jihoo2002 commented on src/main/java/com/keodam/keodam_backend/app/user/service/UserService.java:
```diff
@@ -0,0 +1,105 @@
+
+    private final UserRepository userRepository;
+//    private final FileStorageService fileStorageService;
+
```
`FileStorageService`가 현재 사용되지 않아 주석 처리된 것 같은데, 혹시 이후 사용 계획이 있을까요?  
사용되지 않는다면 정리하고 푸시해도 좋을 것 같습니다!

haeunsong replied:
프로필 사진 업로드 기능을 구현하려면 S3 를 따로 만져야하나 싶어서 두었는데 우선은 제거해야겠네용!

jihoo2002 commented on src/main/java/com/keodam/keodam_backend/app/user/service/UserService.java:
```diff
@@ -0,0 +1,105 @@
+    public UserResponseDto updateRole(User user, RoleType roleType) {
+        if (!roleType.equals("MENTOR") && !roleType.equals("MENTEE")) {
+            throw new IllegalArgumentException("잘못된 역할 선택입니다.");
+        }
```
제 생각에는 문자열 비교방식은 하드코딩이라 `.equals` 말고 다른 방식으로 비교할 수도 있을거 같은데
하은님 의견은 어떠신 지 궁금해요
기존에 있는 `RoleType` 열거형 클래스를 활용해보면 어떨까요?

```java
//예시
 if (roleType != RoleType.MENTOR && roleType != RoleType.MENTEE) {

```

jihoo2002 replied:
>  throw new IllegalArgumentException("잘못된 역할 선택입니다.");

추가로 예외처리 메세지도 `ErrorStatus`에서 공통으로 관리하면 좋을거 같아요

haeunsong replied:
더 좋은 방법인 것 같습니다, 반영해볼게용

jihoo2002 commented on src/main/java/com/keodam/keodam_backend/app/user/dto/req/RoleRequestDto.java:
```diff
@@ -0,0 +1,7 @@
+
+public record RoleRequestDto (RoleType roleType) {
+
+}
```
개행이랑 파일 끝 개행에 신경써주세욥 !
아래 자료는 저도 공유받은 아티클입니당
[파일 끝에 개행이 필요한 이유](https://hyeon9mak.github.io/github-no-newline-at-a-end-of-file/)

jihoo2002 commented on src/main/java/com/keodam/keodam_backend/app/user/dto/res/SignupStatusResponseDto.java:
```diff
@@ -0,0 +1,9 @@
+        boolean hasRole
+){
+
+}
```
여기두!

jihoo2002 commented on src/main/java/com/keodam/keodam_backend/app/user/controller/UserController.java:
```diff
@@ -0,0 +1,93 @@
+@RequestMapping("/user")
+@RequiredArgsConstructor
+public class UserController {
+
```
스웨거 `UI`에서 `API`를 문서화할 수 있도록 스웨거 아노테이션을 추가해주세요!

jihoo2002 commented on src/main/java/com/keodam/keodam_backend/app/user/controller/UserController.java:
```diff
@@ -0,0 +1,93 @@
+
+        String oauthId = oAuth2User.getName();
+        User user = userService.findByOAuthId(oauthId)
+                .orElseThrow(() -> new RuntimeException("사용자를 찾을 수 없습니다."));
```
여기서는 `ErrorStatus` 클래스를 사용하지 않고 직접 명시한 이유가 있을까요?

haeunsong replied:
오류 관련된 걸 나중에 일괄적으로 수정해야겠다 해놓고 안했네요,, 전체적으로 수정해야될 것 같습니당

jihoo2002 reviewed (commented):
고생 많으셨습니다~~~😎😎! 
저도 이번에 두 번째로 코드 리뷰를 해봤는데, 많이 미숙한 부분이 있었던 것 같아요. 
하은님 방향이 더 맞을 수 있으니 적절하게 받아들이시고, 화이팅입니다!🔥

jihoo2002 commented on src/main/java/com/keodam/keodam_backend/app/user/controller/UserController.java:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    @GetMapping("/nickname/check")
+    public ResponseEntity<ApiResponse<String>> checkNickname(@RequestParam String nickname) {
```
저는 `ApiResponse`만 사용해서 응답처리 했는데 
보통 `ResponseEntity`와 `ApiResponse`를 같이 공통 응답 처리로 사용하는 게 보편적으로 사용하는 방식인건가요?
궁금해서 질문 드립니다 !

haeunsong replied:
예를 들어 ApiResponse 만 사용하면, 닉네임이 비어있는 경우에 
```json
{
  "isSuccess": false,
  "code": "400",
  "message": "닉네임이 비어있습니다.",
  "result": null
}
```
이런식으로 나가겠지만, HTTP 상태는 200 OK 로 찍힙니다. 왜냐하면 code 는 우리가 직접 정의해준 부분이고, ApiResponse.onFailure("400", "닉네임이 비어있습니다.", null); 는 정상리턴으로 처리되니까요.

하지만, ResponseEntity 와 함께 사용하면 정의한 400이 아닌, HTTP status 가 400으로 나가기 때문에 보다 정확한 오류를 프론트에 제공할 수 있다고 이해하고 있습니다.

junseok0304 commented on src/main/java/com/keodam/keodam_backend/app/user/controller/UserController.java:
```diff
@@ -0,0 +1,93 @@
+
+@RestController
+@RequestMapping("/user")
+@RequiredArgsConstructor
```
팀내 어노테이션 api 경로 규정에 맞도록 
api/user/ 또는
api/mypage/ 뒤에 맵핑하는 것을 제안합니다.

haeunsong replied:
넵! 반영하겠습니다.

junseok0304 commented on src/main/java/com/keodam/keodam_backend/app/user/controller/UserController.java:
```diff
@@ -0,0 +1,93 @@
+
+    @GetMapping("/nickname/check")
+    public ResponseEntity<ApiResponse<String>> checkNickname(@RequestParam String nickname) {
+        if (nickname == null || nickname.trim().isEmpty()) {
```
닉네임 검증이 너무 빈약한 것 같습니다
팀 피그마에 정의되어있는
닉네임 관련 발생시 예외문구 출력 목록을 가져와봤습니다.

⚠ 특수문자는 닉네임에 포함될 수 없어요
⚠ 중복된 닉네임이에요
⚠ 최초 설정 후 한 번만 변경 가능해요
(30일 이후 변경 가능) 
⚠ 닉네임에 이름은 포함할 수 없어요
⚠ 닉네임에 비속어는 포함할 수 없어요

haeunsong replied:
수정해보겠습니다!

junseok0304 commented on src/main/java/com/keodam/keodam_backend/app/user/domain/User.java:
```diff
@@ -29,7 +30,7 @@ public class User{
     private RoleType roleType;
     @Column(name="oauth_id")
-    private String oAuthId;
+    private String oauthId;
```
변경하신 이유가 있나요?
혹시 해당 부분 변경시 다른 사람이 작성한 코드에 sideEffect는 없었나요?

haeunsong replied:
스프링 데이터 jpa 를 사용할 때 필드명을 oAuthId  로 하면 제대로 인식을 못하더군요.. 말씀드리려고 했는데 깜빡했네요... 그리고 oauth 라는 단어 자체가 o + auth 보다는, 'oauth'라는 하나의 고유한 기술 용어이기 때문에 oauthId 가 더 적합한 것 같습니다!

승현님 코드에서 바꿔야할 부분이 있었는데 단톡방에 여쭤봐야할 것 같습니다.

junseok0304 commented on src/main/java/com/keodam/keodam_backend/app/user/service/UserService.java:
```diff
@@ -0,0 +1,105 @@
+                user.getProfileUrl() != null,
+                user.getRoleType() != null && !user.getRoleType().equals("GUEST")
+        );
+    }
```
회원가입 상태를 확인한다는 아이디어가 좋은 것 같습니다

junseok0304 commented on src/main/java/com/keodam/keodam_backend/global/config/SecurityConfig.java:
```diff
@@ -33,6 +35,9 @@ public class SecurityConfig {
+
+
+
     @Bean
```
공백을 추가하신 이유가 뭔가요?

haeunsong replied:
전체적으로 공백 신경을 안쓰고 커밋해버렸네요. 수정해보겠습니다.

junseok0304 commented on src/test/java/com/keodam/keodam_backend/app/user/controller/UserControllerTest.java:
테스트 코드 작성 좋아요 굳

junseok0304 commented on src/main/java/com/keodam/keodam_backend/app/user/dto/req/NicknameRequestDto.java:
req, res같은 패키지명은 지양하는것이 좋을 것 같습니다. 
이미 클래스명에서 요청인지 응답인지 나와있어서 
dto패키지 내에만 담아두어도 무방할 것 같아요,

junseok0304 commented on src/main/java/com/keodam/keodam_backend/app/user/service/UserService.java:
```diff
@@ -0,0 +1,105 @@
+                .hasProfileImage(user.getProfileUrl() != null)
+                .hasRole(user.getRoleType() != null && !user.getRoleType().equals("GUEST"))
+                .build();
+    }
```
Dto 코드가 서비스 코드 내부에 있는건가요?
이유가 있을까요?

haeunsong replied:
어디에 두어야할지 잘 모르겠어서 우선 여기에다가 두었는데 혹시 어디에다가 두는게 좋을까요...!

junseok0304 replied:
dto 파일로 분기해서 dto패키지 로 관리하는게 좋을 것 같습니다.
저도 처음에는 dto의 의미를 잘 몰랐는데 응답시 순서를 규정하는데에도 역할을 한다고 최근에 알게되었습니다
ex) accessToken과 RefreshToken을 반환하는 api에서 dto가 없다면, 일반적으로는 두 객체의 반환 순서가 랜덤함.
따라서 이를 해결하기 위한 가장 간편한 방법은 dto로 담아서 보내는 것

junseok0304 commented on src/main/java/com/keodam/keodam_backend/app/user/dto/req/RoleRequestDto.java:
recode는 몰랐는데 새로 배워갑니다.

junseok0304 commented on src/main/java/com/keodam/keodam_backend/app/user/service/UserService.java:
```diff
@@ -0,0 +1,105 @@
+    @Transactional
+    public UserResponseDto updateRole(User user, RoleType roleType) {
+        if (!roleType.equals("MENTOR") && !roleType.equals("MENTEE")) {
+            throw new IllegalArgumentException("잘못된 역할 선택입니다.");
```
서비스 코드의 예외처리중 필요한 부분은 항상 리스폰스 코드로 응답하는게 좋을 것 같습니다.

junseok0304 commented on src/main/java/com/keodam/keodam_backend/app/user/domain/User.java:
```diff
@@ -43,18 +44,44 @@ public class User{
+                profileUrl != null ? profileUrl : this.profileUrl,
+                roleType != null ? roleType : this.roleType
+        );
+    }
```
삼항연산자가 무슨 기능을 의미하는지 척보고는 잘 모르겠어요.
null이 아닌 파라미터만 반영해서 선택적으로 필드를 업데이트하는 방식 으로 보이는데

JPA에서는 동일한 @Id를 가진 새 Entity 객체를 만들고 저장하면 문제가 생길 수 있습니다.
Detached entity passed to persist 같은 오류 발생 가능성이 있습니다.

다음과 같은 방식을 제안합니다.
public void update(String nickname, String profileUrl, RoleType roleType) {
    if (nickname != null) this.nickname = nickname;
    if (profileUrl != null) this.profileUrl = profileUrl;
    if (roleType != null) this.roleType = roleType;
}

메서드를 직접 업데이트 해주고, 기존 객체를 수정하는 방향으로

haeunsong replied:
새로 더 배워갑니다. 감사합니다, 반영해보겠습니다!

junseok0304 reviewed (commented):
잘만드셨는데 몇 가지 수정사항에 대해서 코멘트 남겼습니다!
확인해보시고 rc(recheck)주세요!

haeunsong committed:
Refactor(#8): 코드리뷰 반영하여 리팩토링

haeunsong committed:
Feat(#8): 닉네임 검증로직 추가

- 특수문자는 닉네임에 포함될 수 없어요
- 중복된 닉네임이에요
- 최초 설정 후 한 번만 변경 가능해요
  (30일 이후 변경 가능)
- 닉네임에 비속어는 포함할 수 없어요

haeunsong committed:
Merge remote-tracking branch 'origin/develop' into feat/#8

# Conflicts:
#	build.gradle
#	src/main/java/com/keodam/keodam_backend/app/domain/User.java
#	src/main/java/com/keodam/keodam_backend/app/user/repository/UserRepository.java
#	src/main/java/com/keodam/keodam_backend/global/code/status/ErrorStatus.java
#	src/main/java/com/keodam/keodam_backend/global/config/SecurityConfig.java

haeunsong committed:
Fix(#8) : 충돌해결 #8

haeunsong committed:
Feat(#8) : swagger 설정 추가 #8

haeunsong committed:
Refactor(#8) : 불필요한 개행 삭제 및 에러 처리 리팩토링 #8
